README.md
# myShell
Developed a working POSIX compatible shell with a subset of
feature support of default shell. Apart from basic working,
required features implemented are listed below.

Syntax
1. IO redirection with ‘>>’ and ‘>’ will be done for one source and
one destination only.
2. For ‘|’, a minimum 3 programs must be in pipe. E.g. 
cat file1.txt | sort | uniq > save.txt
3. Prompt must change for distinguishing users($)@hostname and
root(#)@hostname and also while searching in extended features.Implementation Requirements

Shell Internals
1. Program is a input driven independent executable, no
menu options are provided. It can take n number of strings
in the format of a command following with its parameters itself.
2. For interactive processes it is a waiting process.

3.It maintain a configuration file which your program reads
on startup and sets the environment accordingly. Like our own
“.bashrc”.
5. Support for these initialization variable are provided -
PATH, HOME, USER, HOSTNAME, PS1​
6. Association of ~ with HOME variable.
7. Look of prompt via PS1​ is to be handled.
8. Alias support for commands. E.g.alias ll='ls -l'.
9. Handle support for - $
​ $, $?​ Similar to shell.

Extended Features :-
1. Recording as script command, where a user can initiate recording
of all input and output to a file, this happens in the
background. It will be invoked as record start  and record stop.
2. History as buffer - 
searchable via TRIE
. This is a list of all
commands executed ever. This can have a cap limit which can be
configured via “.myrc” file in ​
HISTSIZE variable​ . It will be
invoked as history.
3. TRIE data structure for tab completion support is needed. List of
directories to be traversed on startup which are mentioned in
PATH variable for populating all command names into the
structure.
4. Set default applications via .myrc file. This requires extension
matching/mapping to your preferred application. Handle its I/O
and parameter passing on executing.
a. E.g. MEDIA=[‘vlc’, ‘vlc path’,’.mp4’] in config file.
So on running - open same.mp4 , it must invoke vlc.
# !! Mac users be careful for collision of open
keyword in your shell.


history.txt
ramlsls
cat history.txt
rajesh
mahesh
suresh
dinesh
cat history.txt
clear
ls | grep "s"
ls | wc
cd
exit
clear
ls
ls | grep a.out
ls
ls | grep .cpp
cat test.cpp | sort | uniq | wc 
ls
cat history.txt
clear
ls
ls | grep .cpp
ls | grep .cpp > file.txt
ls
cat file.txt
ls | grep .cpp > file.txt
cat file.txt
ls | sort | grep .h > file.txt
cat file.txt
ls
ls  | sort | grep .h > file1.txt
cat file1.txt
CLEAR
clear
exit
su -yash
su - yash
 ls 
     ls
ls
alias sh = " file.txt "
cat sh
alias sh = " file.txt "
cat sh
alias sh = " file.txt "
alias sh = " file.txt "
alias sh = " file.txt "
alias sh = " file.txt "
ls
ls
cat myShell.txt
cat myShell.cpp
clear
alias sh = " file.txt "
alias sh = " file.txt "
alias sh = " file.txt "
alias sh = " file.txt "
alias sh = " ****** "
clear
alias sh = " file.txt "
clear
alias sh = " file.txt "
cat file.txt | ram | jay 
alias sh = " file.txt "
cat file.yx | ram | ja |kjskj |jd
ls
cat file | sjn | ishd | sj
jsj | sj |
clear
ls 
ls
ls
 ls
alias sh = " file.txt "
alias sh = " file.txt "
alias sh = " file.txt "
cat | hh |dh |jd
ls
ls 
ls
ls
alias sh = " file.txt"
alias sh = " file.txt "
alias sh = " file.txt "
alias shp = " ls -l "
shp
clear
ls
alias sh = " ls -l "
sh
alias
alias ll = " ls -l "
ll ..
ll
alias ll = " ls "
ls
ll
ls
ls
alias ll = " ls "
ll
ls
alias ll = " ls -l "
ll
ls
ls
alias sh = " cat "
sh history.txt
cat history.txt
clear
ls
ls
clear
ls 
alias ll = " ls -l "
ll
ls
alias f = " history.txt "
cat f
cat f | grep cat
ls
g++ myShell.cpp
./a.out
cd ..
ls
g++ Alias_handling.cpp
./a.out
clear
sleep 5
sleep 5 &
script file.txt
script file.txt
ls
script file.txt
ls
script file.txt
ls
ls
lskd
skm
ls
sk
script file.txt
ls
script file.txt
ls
ls -l
ls
ls -l
ls
echo cool
ls -l
clear
script file.txt
ls
ls -l
ls
ls -l | sort | uniq > file2.txt
script file
ls
echo cool
script file
ls
ls
echo cool
script file.txt
ls
ls -l | sort 
cat myShell.cpp
ls -l
script
script
ls
script
script 
ls
ls -l
ls -l | sort
ls | sort
cat myShell.cpp | grep include | sort
script
ls 
ls | sort
script
ls
ls | sort
script 
ls
ls | sort
ls | sort
script
ls | sort
script
ls | sort
script
ls | sort
script
ls | sort
script 
ls | sort
script
ls 
ls | sort
script
ls | sort

   
    
 ls
          ls

ls
slslslssls
                    

script
ls
ls -l | sort
cat >> default_file.txt

cat myShell.cpp >> default_file.txt
cat myShell.cpp > default_fiel.txt
ls -l | sort > default_file.txt
clear
ls
pwd > default_file.txt
ls > default_file.txt
clear
exit
cat > filetest.txt

ls
kdj
exit
hjh
ls
ls -l
open myShell.cpp
open file.txt
open myShell.txt
ls
open myShell.cpp
open history.txt
open myShell.cpp
open myShell.cpp
open myShell.cpp
open myShell.cpp
open history.txt
open myShell.cpp
open history.txt
open history.txt
ls
ls
clear
ls
ls
clear
open history.txt
ls
ls
ls
ls
ls
ls
open history.txt
ls
ls
open history.txt
ls
clear
mkdmsd
ls
ls
ls
ls
ls -l
open history.txt
ls
ls -l
ls
open history.txt
ls
ls
ls
ls
ls
open history.txt
ls
ls
ls
ls
ls
ls
clear
ls
open history.txt
ls
ls
ls
cs
js
sk
alias ll = " ls -l "
ll
ll | sort
ll | sort | grep cpp
open history.txt
open myShell.c
cd /.
ls
cd ..
cd /.
ls
ls
ll
cd /
ls
cd ..
cd ..
cd ..
cd ..
cd ..
cd ..
ls
open history.txt
ls
cd ..

ls
cd ..
                  
                                     ls
                 
       
        
    
                     
clear
cd ..
alias grepe = " grep --color=auto "
ls
ls -l
ls | grepe cpp
alias la = " ls -l "
la
alias grepe = " grep --color=auto "
grep
grep
ls | grep out
alias g = "sort"
g
alias g = " sort "
g
clear
grepe = " grep --color=auto "
alias grepe = " grep --color=auto "
grepe
alias grepe = " grep --color=auto "
ls -l | grepe cpp
alias grepe = " grep --color=auto "
ls -l | grepe
alias grepe = " grep --color=auto "
grepe
alias grepe = " grep --color=auto "
alias grepe = " grep --color=auto "
alias grepe = " grep --color=auto "
alias grepe = " grep --color=auto "
alias grepe = " grep --color=auto "
alias grepe = " grep --color=auto "
alias grepe = " grep --color=auto "
alias grepe = " grep --color=auto "
grepe


alias grepe = " grep --color=auto "
grepe
ls -l | grepe
ls -l | grepe out
clear
cd 
exit
record start file1.txt
ls
ls -l
record start 
ls -l
top
cd ~
cd ..
cd ../..
record start file1.txt
ls
ls -l
record start file.txt
ls
ls -l
top
top
top

ld
ls
top
record start file.txt
top
ls
record start file.txt
top

ls
top
ls
cd ..
record start
ls
ls -l
ls -l
ls -l
record start 
ls
ls
top
q
ls
ls -l
q
ls -l
q
q
q
record start
ls
ls -l
ps
top 
ls 
ps
ls -l
record start 
ls
top 
ls -l
record start
ls
record stop
ls
ps
ls
ls
ls
ls
ls
ls
ls
ls
sls
sls
ls
sk
sks
clear
ls
ls
cls
smnjask
sdnk
njsnd
jsk
ks
ddk
history
ls 
ls
ls
mlsdl
sdoska
sldml
mlsamd
msldm
history
history
ls
yeah
history
ls
ll
gc
ls
ll
ls
ll
clear
yeah
ll
ls
history
ls
lq
l[
lp
ls
ls
ll
ls
[
ls
vh
jbjb
kjbkj
jbj
jbj
jbj
jb
mm nm
hjvhvv
jjb
jhvhj
 n n 
hhvh
which vlc
exit
history
ls
history
ls
ls -l
history
echo ls
echo mkmsdkmcas
echo $$
record start
ls
ll





exit








record start
ll
record start
ls -l
ll
ljj
ll

ll



ls
record start
ll
ll
ll
ls
ls
ll
lp
msk
ls
ls -l
pp
ls
ll
ll
ls
ll
lp
ko
ls
ls | sort
ls
ll
lp
uihih
maldsm
s,dl
ds
s
ls
ll
dksd
sd
lp
l
ls
echo $ram
ls
echo $$
echo $?
ls
ls
echo $$
proc
ps
ls 
echo $$
echo $?
echo $PATH
echo $EDITOR
echo $EDITOR
echo $$
echo $HOME
echo $HISTSIZE
echo $EDITOR
echo $HOME
ls
llp
llp
lskd
sksjd
skkso
sasko
saskk
sls
sls
sls
sls
sls
lcld
slsld
dldls
clear
ls
ll
ji
kmmk
,oll
l,l
ooko
l,,l
lnnj
kkmk
ll
ommo
mkk
uuyb
jnjk
hhbjbhhj
jjnjkb
jjn
jjk
ls
ls
ll
lsl
lsls
cks
mksc
lsc
iiu
njknjk
jjkb
 jkbj
iuuiui\
nm
kk
ls
ook
k
ls l
kmksm
ls
ls

sksk
ksks
ls
ls
ls
ls
ls
csc
cs
sckmd
scac
echo $$
echo $?
ls
ls -l
ls
ls
ll
ls
ll
ls
ll
echo $?
ls
ll
lp
ls
echo $PATH
export PATH la
echo $PATH
ls
ps ~
cd ~
cd /
ls
ls
ls
ls
ls
ls
ls
cd ..
ls
ls
ls
ls
ls -l | sort | grep cpp > file1.txt
ls -l | sort | grep > file1.txt
ls -l | sort | grep cpp > file1.txt
ls -l | sort | grep cpp >> file1.txt
alias s = " sort "

Init.h
#include <stdio.h>
#include <fcntl.h> 
#include <sys/utsname.h>
#include <pwd.h>
#include <stdlib.h>

#include<bits/stdc++.h>
#include <stdio.h> 
#include <sys/types.h> 
#include <unistd.h>  

#include <errno.h>   
#include <sys/wait.h>
#include<iostream> 

using namespace std;


std::vector<char *> file_type;
std::vector<char *> app_path;
std::vector<char *> app_name;



void handle_open(char * file)
{
	char *argv[10];
	char ** temp=argv;

 	char file_copy[100];
 	strcpy(file_copy,file);
 	char* token = strtok(file, ".");

	while (token != NULL) {  
		*temp=token;
		*temp++;
		token = strtok(NULL, " "); 
	} 

	*temp=NULL;
	*temp--;

	int i;
	for(i=0; i<file_type.size();i++)
	{
		if(!strcmp(file_type[i],*temp))
			break;
	}
	if(i>=file_type.size())
	{
		cout<<"Error:No associated application\n";
		return;
	}

	if(!fork()){
		close(1);
    	execlp(app_name[i], app_path[i], file_copy, NULL);
	}

}

void handle_$(char * command)
{
	command++;
	

	if(!strcmp(command, "$"))
	{
	  	cout<<getpid()<<"\n";
	  	return;
	}

	if(!strcmp(command, "?"))
	{
	  	cout<<last_proc_status<<"\n";
	  	return;
	}
	//cout<<command;
	if(getenv(command))
		printf("%s\n", getenv(command));

}




void init()
{

	// file_type.push_back("txt");
	// app_name.push_back("gedit");
	// app_path.push_back("/usr/bin/gedit");

	alias_map.insert({"~", getenv("HOME")} );
  	int fd, sz; 
  	char *c = (char *) calloc(1000, sizeof(char)); 

  	uid_t uid=getuid();

  	struct passwd * my_passwd;
	my_passwd = getpwuid(uid);

	struct utsname name;
	if (uname(&name)==-1) {
	    //die("%s",strerror(errno));
	    printf("Error in getting hostname");
	}
	const char* hostname=name.nodename;
  
  	fd = open("myrc.txt", O_RDONLY); 
  	if (fd < 0) { perror("Error: unable to load myrc"); exit(1); } 
   	sz = read(fd, c, 1000); 
	c[sz] = '\0'; 

	
	char* token = strtok(c, "\n");

	setenv("PATH",token,1); 
	
	setenv("USER",my_passwd->pw_name,1);

	setenv("HOME",my_passwd->pw_dir,1);// does overwrite

	setenv("HOSTNAME",hostname,1);

	char ps1[100]="";

	sprintf(ps1,"%s%s%s%s",getenv("USER"),"@",getenv("HOSTNAME"),":");
	
	setenv("PS1",ps1,1);

	token=strtok(NULL,"\n");

	setenv("HISTSIZE",token,1);

	token = strtok(NULL, "\n"); 

	while(token != NULL)
	{
		strcat(token,"");
		//cout<<token;
		//token = strtok(NULL, "\n"); 
		file_type.push_back(token);
		token = strtok(NULL, "\n"); 
		app_name.push_back(token);
		token = strtok(NULL, "\n"); 
		app_path.push_back(token);

		token = strtok(NULL, "\n"); 


	}

	// printf("PATH : %s\n", getenv("PATH"));
	// printf("PATH : %s\n", getenv("USER"));
	// printf("PATH : %s\n", getenv("PS1"));	

	// for(int i=0;i<file_type.size();i++)
	// {
	// 	cout<<"\n"<<file_type[i]<<" "<<app_name[i]<<" "<<app_path[i]<<"\n";
	// }


	close(fd);

	//return 0;
}

Alias_handling.h
#include<bits/stdc++.h>
#include <stdio.h> 


using namespace std;

map<string, char *> alias_map;
map<string, char *>::iterator itr; 

void parseCommand_alias(char * command,char **argv)
{
		char* token = strtok(command, " "); 

	  	char ** temp=argv;
	  
		// Keep printing tokens while one of the 
		// delimiters present in str[]. 
		while (token != NULL) { 
			 
	      	*temp=token;
	      	*temp++;
			token = strtok(NULL, " "); 
		} 

		*temp=NULL;

		//cout<<command<<"\n";

}

void check_for_alias(char * command)
{
	//cout<<"i mhere\n";
	char * argv[50];
	parseCommand_alias(command,argv);
	
	char ** temp=argv;
	//cout<<"i m here \n";

	while(*temp)
	{
		//printf("%s", *temp);

		itr=alias_map.find(*temp);

		if(itr!=alias_map.end())
		{
			*temp=itr->second;
			//cout<<"I m in";
		}

		*temp++;
	}

	string value="";

	temp=argv;

	while(*temp )
	{
		//cout<<*temp<<"i m \n";
		value=value+*temp;
		value=value+" ";
		*temp++;
	}

	//cout<<" \nvalue ="<<value;

	//char * p =(char *)malloc(value.length() * sizeof(char*) ); 
  	//cout<<"\n" << sizeof(p)<<" "<< value.length();
  	char p[1024];
  	int i;
    for (i = 0; i < value.length()+1; i++) { 
        p[i] = value[i]; 
        //cout << p[i]; 
    } 
    //p[i]=NULL;
    //cout<<"\n p = "<<p;
    strcpy(command,p);
    //free(p);
}


char p[256];

void insert_alias_map(char ** argv)
{
	string key=argv[1];
	char ** temp =argv+4;
	string value="";
	//cout<<argv[4];
	while(*(temp+1) )
	{
		value=value+*temp;
		value=value+" ";
		*temp++;
	}

	cout<<key<<" \t"<<value;

     
  	//cout<<"\n" << sizeof(p)<<" "<< value.length();
    for (int i = 0; i <= value.size(); i++) { 
        p[i] = value[i]; 
        //cout << p[i]; 
    } 
   // cout<<" value of p ="<<p;
    pair <string, char*> PAIR1 ; 
  
    PAIR1.first = key; 
    PAIR1.second = p ;
    //cout<<PAIR1.second; 
	alias_map.insert(PAIR1);
}




Alias_handling.cpp
#include<bits/stdc++.h>
#include <stdio.h> 


using namespace std;

map<string, char *> alias_map;
map<string, char *>::iterator itr; 

void parseCommand(char * command,char **argv)
{
		char* token = strtok(command, " "); 

	  	char ** temp=argv;
	  
		// Keep printing tokens while one of the 
		// delimiters present in str[]. 
		while (token != NULL) { 
			 
	      	*temp=token;
	      	*temp++;
			token = strtok(NULL, " "); 
		} 

		*temp=NULL;

		//cout<<command<<"\n";

}

void check_for_alias(char * command)
{
	//cout<<"i mhere\n";
	char * argv[50];
	parseCommand(command,argv);
	
	char ** temp=argv;
	//cout<<"i m here \n";

	while(*temp)
	{
		//printf("%s", *temp);

		itr=alias_map.find(*temp);

		if(itr!=alias_map.end())
		{
			*temp=itr->second;
			//cout<<"I m in";
		}

		*temp++;
	}



	string value="";

	temp=argv;

	while(*temp )
	{
		//cout<<*temp<<"i m \n";
		value=value+*temp;
		value=value+" ";
		*temp++;
	}

	//cout<<" \nvalue ="<<value;

	char * p =(char *)malloc(value.length() * sizeof(char) ); 
  	//cout<<"\n" << sizeof(p)<<" "<< value.length();
    for (int i = 0; i < value.length(); i++) { 
        p[i] = value[i]; 
        //cout << p[i]; 
    } 
    //cout<<"\n p = "<<p;
    strcpy(command,p);
}

void init_alias()
{
	char * ll="ls -alF";
	char * la="ls -A";
	char * l="ls -CF";
	string temp="ll";
    alias_map.insert(pair<string, char *>(temp, "shrayans")); 
    alias_map.insert(pair<string, char *>("la", la)); 
    alias_map.insert(pair<string, char *>(l, l)); 
    // //cout<<"alisas updated \n";
    //  for (itr = alias_map.begin(); itr != alias_map.end(); ++itr) { 
    //     cout <<"\n" << itr->first <<" "
    //           << itr->second << '\n'; 
    // } 
}

void insert_alias_map(char ** argv)
{
	string key=argv[1];
	char ** temp =argv+4;
	string value="";
	//cout<<argv[4];
	while(*(temp+1) )
	{
		value=value+*temp;
		value=value+" ";
		*temp++;
	}

	//cout<<key<<" "<<value;

    char * p =(char *)malloc(value.length() * sizeof(char) ); 
  	//cout<<"\n" << sizeof(p)<<" "<< value.length();
    for (int i = 0; i < value.length(); i++) { 
        p[i] = value[i]; 
        //cout << p[i]; 
    } 
    //cout<<" value of p ="<<p;
    
	alias_map.insert(pair<string, char *>(key, p) );
}


int main()
{

	for (int i = 0; i < 1; i++)
	{
		char c[100];
		cin.getline(c,100);
		char * argv[50];
		//cout<<c<<" " ;

		parseCommand(c,argv);
		//cout<<argv[0]<<" ";
		if(!strcmp(argv[0], "alias"))
		{
			//cout<<"i m in";
			insert_alias_map(argv);
		}
	}

	init_alias();
	cout<<"enter command\n";
	char command[100];
	cin.getline(command,100);

	//char *argv[50];
	check_for_alias(command);


	// int i=0;
	// while(argv[i])
	// 	cout<<argv[i++]<<" " ;
	cout<<command<<"\n";
}

History_maintain.h
#include <stdio.h>
#include <stdlib.h>
#include<bits/stdc++.h>

#define BUFFER_SIZE 1000

#define FileNAme "history.txt"

vector<char * > history_vector;

void update_history_vector(char * command)
{
   char *s =getenv("HISTSIZE");
   char *end;
   //cout<<" size *** "<< history_vector.size()<<" " << command<<"\n";

   long n = strtol(s,&end,10);

   if(n<=history_vector.size())
   {
        history_vector.erase(history_vector.begin());
        history_vector.push_back(command);
        return;
   }
   char *t=new char[100];
   strcpy(t,command);
   //cout<<"Yes";

   history_vector.push_back(t);

}


void  update_History(char * command)
{

    FILE *fPtr;
    char filePath[100];

    fPtr = fopen(FileNAme, "a");

    if (fPtr == NULL)
    {
        printf("Error in updating history");
    }

    fputs(command, fPtr);
    fputs("\n",fPtr);

    fclose(fPtr);

    update_history_vector(command);
}



void print_history()
{
    for(int i=0 ;i<history_vector.size();i++)
    {
        cout<<history_vector[i]<<"\n";
    }
}


IO_handle_header.h
#include<bits/stdc++.h>
#include <stdio.h> 
#include <sys/types.h> 
#include <unistd.h>  
#include <stdlib.h> 
#include <errno.h>   
#include <sys/wait.h> 
#include <unistd.h>
#include<fcntl.h>

using namespace std;

int last_proc_status=0;

bool script_flag=false;
char  script_path[256];

bool check_single_arrow=false;

std::vector<char *> v;
std::vector<char *> vector_IO;

void parseCommand_IO(char * command,char **argv)
{
		char* token = strtok(command, " "); 

	  	char ** temp=argv;
	  
		// Keep printing tokens while one of the 
		// delimiters present in str[]. 
		while (token != NULL) { 
			//printf("%s\n", token); 
	      	*temp=token;
	      	*temp++;
			token = strtok(NULL, " "); 
		} 

		*temp=NULL;
}



void exec_for_script_io(char * command)
{
	int fd = open(script_path, O_CREAT | O_RDWR | O_APPEND, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
	int sz = write(fd, command, strlen(command));
	sz = write(fd, "\n", strlen("\n"));  
	int x=dup(1);

	if(!fork())
	{
		pid_t pid;
		char * argv[50];
	  	parseCommand_IO(command,argv);

	  	//cout<<argv[0]<<"\t";//<<argv[1];
	  	dup2(fd,1);

	     pid=fork();

	     if(pid==0)
	     {
	     	if(execvp(argv[0], argv) < 0)    
                printf("Error: command not found\n");
	     }
	     else{
	     	wait(NULL);
	     	abort();
	     }
	     
	}
	else
	{
		wait(NULL);
		dup2(x,1);
		close(fd);
	}
}


bool checkIO(char * command)
{
	while(*command)
	{
		if(*command==124 || *command=='>')
			return true;
		command++;
	}
	return false;
}

void split_by_pipe(char * command)
{


		//cout<<"IM";
	  	char *end_str;
		char* token_pipe = strtok_r(command, "|", &end_str); 
		//cout<<token_pipe;

		while(token_pipe!=NULL)
		{
			v.push_back(token_pipe);
			//cout<<token_pipe;
			token_pipe = strtok_r(NULL, "|",&end_str);
		}

		// for(int i=0 ;i<v.size();i++)
		// 	cout<<"\n i "<<i<<v[i];
		//cout<<"\n";

}

void split_by_arrow(char * command)
{
		char *end_str;
		char* token_pipe = strtok_r(command, ">", &end_str); 
		//cout<<token_pipe;

		while(token_pipe!=NULL)
		{
			vector_IO.push_back(token_pipe);
			//cout<<token_pipe;
			token_pipe = strtok_r(NULL, ">",&end_str);
		}
		// for(int i=0 ;i<vector_IO.size();i++)
		// 	cout<<"\n i "<<i<<vector_IO[i];
		// cout<<"\n";

}



bool check_arrow_string(char * command)
{	int count=0;
	bool return_val=false;
	while(*command)
	{
		if(*command=='>'){
			 
			return_val=true;
			count++;
		}

		command++;
	}

	if(count==1)
		check_single_arrow=true;

	return return_val;	
}


void redirecting_to_file( char *c)
{

	//cout<<c;

	split_by_arrow(c);

	char * argv[50];
	parseCommand_IO(vector_IO[1],argv);

	int fd;
	if(check_single_arrow){
		fd = open(argv[0], O_CREAT | O_RDWR | O_TRUNC,S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
	}
	else{
		fd = open(argv[0], O_CREAT | O_RDWR | O_APPEND,S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
	}

	if(!fork())
	{
		char * argve[50];
		parseCommand_IO(vector_IO[0],argve);
		dup2(fd,1);
		execvp(argve[0],argve);
		perror("exec");

	}
	//wait(NULL);
	close(fd);
	//cout<<"exit hogya hu m ";
	//exit(0);
	//exit(0);
}

void handleIO( char * command) 
{

	split_by_pipe(command);

	int i;
	char **argv=NULL;


	if(!fork())

	{
		//pipe(fd);	
	    for( i=0; i<v.size()-1; i++)
	    {
	        int pd[2];
	        pipe(pd);

	        if (!fork()) {

	        	char * argv[50];
	        	parseCommand_IO(v[i],argv);
	            dup2(pd[1], 1); // remap output back to parent
	            execvp(argv[0], argv);
	            perror("exec");
	            //abort();
	        }

	        // remap output from previous child to input
	        
	        dup2(pd[0], 0);
	        close(pd[1]);
	    }

	    if( check_arrow_string(v[i]) )
	    {	

	    	redirecting_to_file(v[i]);
	    	abort();
	    }
	    else
	    {	

	  //   	if(script_flag)
			// {	
			// 	exec_for_script_io(v[i]);
			// }
			//cout<<"\nflow is here\n"<<x;
		    char * argve[50];
		    parseCommand_IO(v[i],argve);
		    //wait(NULL);
		    //dup2(fd[1],0)
		    execvp(argve[0], argve);
		    perror("exec");
		    abort();
		}
	}
	else{
		wait(NULL);
		v.clear();
		vector_IO.clear();
	}
	//cout<<"ram";

}



	//handleIO(argv);


	// char** commands[10];

	// int i=0;

	// char **temp = argv;

	// while(temp!=NULL)
	// {
	// 	char * * temp1;
	// 	while(strcmp(*temp, "|"))
	// 	{
	// 		//temp1=**(commands+i);
	// 		*temp1=*temp;
	// 		*temp1++;
	// 		*temp++;
	// 	}
	// 	temp1=NULL;
	// 	commands[i]=temp1;
	// 	temp++;
	// 	i++;
	// }


	// int pfds[2];
 //    pipe(pfds);

 //    pid_t childpid = fork();

 //    if (childpid == 0) {
 //        /* Child */
 //        dup2(pfds[1],1);
 //        close(pfds[0]); 

 //        execlp("ls", "ls",NULL);

 //    } else {
 //        /* Parent */

 //        pid_t retpid;
 //        int  child_stat;
 //        //while ((retpid = waitpid(childpid, &child_stat, 0)) != childpid && retpid != (pid_t) -1);

 //        wait(NULL);

 //        close(pfds[1]); 

 //        char buf[100];
 //        ssize_t bytesread;

 //        int fd = open("beejoutput.txt", O_CREAT | O_RDWR, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
 //        if (fd == -1) {
 //            fprintf(stderr, "Opening of beejoutput.txt failed!\n");
 //            exit(1);
 //        }

 //        /* This part writes to beejoutput.txt */
 //        while ((bytesread = read(pfds[0], buf, 100)) > 0) {
 //            write(fd, buf, bytesread);
 //        }

 //        lseek(fd, (off_t) 0, SEEK_SET);
 //        dup2(fd, 0);
 //        execlp("wc", "wc", "-l", NULL);
 //    }



	// if(fork()==0)
	// {
	// 	int fd=open("ioCheck.txt", O_RDWR | O_CREAT, S_IRUSR | S_IWUSR);
	// 	dup2(fd,1);
	// 	dup2(fd,2);

	// 	close(fd);

	// 	execvp(argv[0], argv);
	// }


myShell.cpp
#include<bits/stdc++.h>
#include <stdio.h> 
#include <sys/types.h> 
#include <unistd.h>  
#include <stdlib.h> 
#include <errno.h>   
#include <sys/wait.h> 
#include "IO_handle_header.h"
#include "History_maintain.h"
#include "Alias_handling.h"
#include "Init.h"

using namespace std;

// bool script_flag=false;
// char  script_path[150]="default_file.txt";

void exec_for_script(char * command)
{
	int fd = open(script_path, O_CREAT | O_RDWR | O_APPEND,S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
	int sz = write(fd, command, strlen(command));
	sz = write(fd, "\n", strlen("\n"));  
	int x=dup(1);

	if(!fork())
	{
		pid_t pid;
		char * argv[50];
	  	parseCommand_IO(command,argv);

	  	//cout<<argv[0]<<"\t";//<<argv[1];
	  	dup2(fd,1);

	  	if(!strcmp(argv[0], "alias"))
		{
			//cout<<"i m in";
			insert_alias_map(argv);
			//continue;
		}


	  	if(!strcmp(argv[0], "cd"))
	  	{	
	  		//cout<<"yes i m in cd\n";
	  		chdir(argv[1]);
	  		//continue;
	  	}

	     pid=fork();

	     if(pid==0)
	     {
	     	if(execvp(argv[0], argv) < 0)    
                printf("Error: command not found\n");
            abort();
	     }
	     else{
	     	wait(NULL);
	     	abort();
	     }
	     
	}
	else
	{
		//wait(NULL);
		dup2(1,x);
		close(fd);
	}
}

void printPromt()
{

  	uid_t uid;/*uid_t is a build in data type, which can hold small numbers*/
    string username;
    uid=getuid();
 //    username=getlogin();

	char pwd [200];
	getcwd(pwd, 100);

	// char hostname[100];
 //    gethostname(hostname, 1024);

 //    string PS1="\n"+username+"@"+hostname+":~"+pwd;

    if(uid=0)
    {
    	cout<<"\n"<<getenv("PS1")<<"# ";
    }
    else
    {
		cout<<"\n"<<getenv("PS1")<<pwd<<"$ ";
	}
}

void parseCommand(char * command,char **argv)
{
		char* token = strtok(command, " "); 

	  	char ** temp=argv;
	  
		// Keep printing tokens while one of the 
		// delimiters present in str[]. 
		while (token != NULL) { 
			//printf("%s\n", token); 
	      	*temp=token;
	      	*temp++;
			token = strtok(NULL, " "); 
		} 

		*temp=NULL;
}

char command_copy[1024];

int main()
{
	init();
	
	while(1)
	{

		//clrscr();
		printPromt();
		
		pid_t pid;
		int status;

	    char command[1024];

	    cin.getline(command,1024);

	    //updating history

	    update_History(command);

	    //check for alias in command
		check_for_alias(command);
		//cout<<"\ncommand : before script-"<<command<<"\n";

		strcpy(command_copy,command);


	  	if( checkIO(command) ){
	  		//cout<<"getio\n";
	  		handleIO(command);
	  		continue;
	  	}


	  	char * argv[50];
	  	parseCommand(command,argv);

		if(argv[0]==NULL)
			continue;
	  	//cout<<argv[0]<<"\t";//<<argv[1];

	  	if(!strcmp(argv[0], "alias"))
		{
			//cout<<"i m in";
			insert_alias_map(argv);
			continue;
		}


	  	if(!strcmp(argv[0], "cd"))
	  	{	
	  		//cout<<"yes i m in cd\n";
	  		chdir(argv[1]);
	  		continue;
	  	}

	  	if(!strcmp(argv[0], "exit"))
	  	{	
	  		exit(0);
	  	}	  	

	  	if(!strcmp(argv[0], "record"))
	  	{	
	  		if(!strcmp(argv[1],"start"))
	  			script_flag=true;
	  		else
	  			script_flag=false;

	  		if(argv[2]!=NULL)
	  			strcpy(script_path,argv[2]);

	  		getcwd(script_path,150);
	  		strcat(script_path,"/record_file.txt");
	  		cout<<"\n"<<script_path;
	  		continue;
	  	}

	  	if(!strcmp(argv[0], "open") && argv[1]!=NULL)
	  	{	
	  		//cout<<"yes i m in cd\n";
	  		handle_open(argv[1]);
	  		continue;
	  	}

	  	if(!strcmp(argv[0], "history"))
	  	{
	  		print_history();
	  		continue;
	  	}

	  	if(!strcmp(argv[0], "echo"))
	  	{
	  		if(*argv[1]=='$')
	  		{
	  			handle_$(argv[1]);
	  			continue;
	  		}
	  		//continue;
	  	}

	  	if(!strcmp(argv[0], "export"))
	  	{
	  		setenv(argv[1],argv[2],1);
	  		continue;
	  	}

	  	if(script_flag)
		{
			exec_for_script(command_copy);
		}
		//cout<<"\ncommand : after script-"<<command_copy<<"\n";

	     pid=fork();

	     if(pid==0)
	     {
	     	if(execvp(argv[0], argv) < 0){    
                printf("Error: Command not found");
                
                exit(1);

            }
	     }
	     else{
	     	waitpid(pid, &status, 0);
	     	last_proc_status=WEXITSTATUS(status);
	     //cout<<"\nLast process status "<<last_proc_status;
	 		}
	}


}
		


//cp /home/shrayans/raj.cpp /home/shrayans/Documents


//https://github.com/shrayans/myShell.git

//https://github.com/IIIT-OS-PG/shell-shrayans